### PR TITLE
Add feature blog for CEL for Admission Control

### PR DESCRIPTION
**Pull request by jpbetz**
PR for feature blog on KEP: https://github.com/kubernetes/enhancements/issues/3488

### CODE REVIEW

**Comment by jpbetz** [comment]
cc @fsm @sftim

**Comment by sftim** [comment]
Thanks

/hold

OK to unhold once this targets _main_ and has a publication date assigned.

**Comment by jpbetz** [comment]
Retargeted against main.

/retest

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+We are working hard to add more features to admission policies and make the
+enhancement easier to use. Try it out, send us your feedback and help us build
+a simpler alternative to admission webhooks!
+
+## How do I get involved?
+
+If you want to get involved in development of admission policies, discuss enhancement
+roadmaps, or report a bug, you can get in touch with developers at
+[SIG-API-Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:150]:
```suggestion
[SIG API Machinery](https://github.com/kubernetes/community/tree/master/sig-api-machinery).
```

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+metadata:
+  name: "replica-limit-production.example.com"
+maxReplicas: 1000
+```
+
+This binding and parameter resource configure deployments in namespaces with the
+`environment` label set to `production` to a max limit of 1000 replicas.
+
+A seperate binding can then be configured to set a lower limit for `test`
+namespaces.
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:134]: Two nits:

---

“separate” (two `e`s, two `a`s)

---

```suggestion
You can then configure a separate binding in order to set a lower limit for namespaces
in the `test` environment.
```

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    name: "replica-limit-production.example.com"
+  matchResources:
+    namespaceSelectors:
+    - key: environment,
+      operator: In,
+      values: ["production"]
+```
+
+```yaml
+apiVersion: rules.example.com/v1
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:123]:
```suggestion
apiVersion: rules.example.com/v1 # defined via a CustomResourceDefinition
```
?

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+For example, the above admission policy can be modified to make it configurable:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+Spec:
+  paramKind:
+    apiVersion: rules.example.com/v1
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:88]:
```suggestion
    apiVersion: rules.example.com/v1 # You also need a CustomResourceDefinition for this API
```
?

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+`environment` label set to `production` to a max limit of 1000 replicas.
+
+A seperate binding can then be configured to set a lower limit for `test`
+namespaces.
+
+I hope this has given you a glimpse of what is possible with validating
+admission policies! There are many features that we have not yet touched on.
+Have a look at the [feature
+documentation](content/en/docs/reference/access-authn-authz/validating-admission-policy.md)
+to learn more.
```
Comment (sftim) [inline content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:140]:
```suggestion
To learn more, read
[FIXME](/docs/reference/access-authn-authz/validating-admission-policy/).
```

Use the title of the page there. However, https://deploy-preview-37810--kubernetes-io-main-staging.netlify.app/docs/reference/access-authn-authz/validating-admission-policy/ doesn't work yet. Is there a PR that needs to land first?

**Review by sftim — COMMENTED** [review]
Thanks both. Here's some feedback.

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+Next bind the policy the the appropriate resources:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "demo-binding-test.example.com"
+spec:
+  policy: "replicalimit-policy.example.com"
```
[the remark this answers is not in the document]
Comment (jpbetz) [reply content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:60]: Good catch. My copy-paste got sloppy.. Fixing.

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md (hunk 4 above)
Comment (jpbetz) [reply content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:88]: Thanks, this helps.

File: content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md (hunk 5 above)
Comment (jpbetz) [reply content/en/blog/_posts/2022-11-29-validating-admission-policies-alpha/index.md:140]: Correct, this requires https://github.com/kubernetes/website/pull/37770

**Comment by jpbetz** [comment]
Thanks @sftim! Feedback applied. This does depend on https://github.com/kubernetes/website/pull/37770.

**Comment by fsmunoz** [comment]
Thank you @jpbetz, tracking updated on our end, will also give it a look during the day/week.

**Comment by fsmunoz** [comment]
Hello, Comms Lead for v1.26 here: this Feature Blog PR is **confirmed for the post-release Feature Blog series**. We will add the publishing date in the next days. Thank you for your work!

**Comment by fsmunoz** [comment]
Hello, Comms Lead for v1.26 here. **The publication order and date for the Feature Blog series has been finalised** and the [tracking board](https://github.com/orgs/kubernetes/projects/98/views/11?sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=22229692&groupedBy%5BcolumnId%5D=15217749&filterQuery=enhancements-status%3ATracked+blog-status%3A%22Ready+for+Review%22) updated (cf. [discussion](https://github.com/kubernetes/sig-release/discussions/2105)).

The publication date for this article is **20-12-2022 (December 20)**. 

Thanks!

**Comment by jpbetz** [comment]
> The publication date for this article is **20-12-2022 (December 20)**.

Great, I've updated the date in this PR to match.

**Comment by sftim** [comment]
/sig api-machinery

Let's get a tech review on this.

**Comment by sftim** [comment]
@jpbetz if this gets LGTM, the repo lets you do a squash and keep LGTM provided you retain the same merge base.

A tip: `git rebase -i "$(git merge-base main HEAD)"`

It's slightly tidier if you do the squash and then we remove the label that would otherwise tell Tide to request one.

**Comment by sftim** [comment]
@jpbetz, do you know who could check technical accuracy on this? Another API Machinery contributor perhaps?

**Comment by fsmunoz** [comment]
Pending technical review, I've asked in #sig-api-machinery for some help in this in order to move it to Ready for Publication.

**Comment by jpbetz** [comment]
@cici37 would you do a quick pass on technical correctness?

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "demo-binding-test.example.com"
+spec:
+  policy: "demo-policy.example.com"
+  matchResources:
+    namespaceSelectors:
```
Comment (cici37) [inline content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:65]: The field name is `namespaceSelector`

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "demo-binding-production.example.com"
+spec:
+  policy: "demo-policy.example.com"
+  paramsRef:
+    name: "demo-params-production.example.com"
+  matchResources:
+    namespaceSelectors:
```
Comment (cici37) [inline content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:122]: `namespaceSelector`

**Review by cici37 — COMMENTED** [review]
overall LGTM from technical side, just couple nits

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md (hunk 7 above)
Comment (jpbetz) [reply content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:65]: Thanks, fixed.

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md (hunk 8 above)
Comment (jpbetz) [reply content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:122]: Thanks, fixed.

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+Language](https://github.com/google/cel-spec) (CEL) to offer a declarative,
+in-process alternative to [validating admission
+webhooks](/docs/reference/access-authn-authz/extensible-admission-controllers/#what-are-admission-webhooks).
+
+CEL was first introduced to Kubernetes for the [Validation rules for
+CustomResourceDefinitions](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules).
+This enhancement expands the use of CEL in Kubernetes to support a far wider
+range of admission use cases.
+
+Admission webhooks are burdonsome to develop. Webhook developers must implement
```
Comment (JimBugwadia) [inline content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:23]: Admission webhooks can be burdensome to develop and operate.

Comment (JimBugwadia) [inline content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:23]: burdensome

**Comment by sftim** [comment]
/hold
OK to unhold once v1.26 is released.

**Comment by sftim** [comment]
/lgtm
/approve

Remember, we can accept tweaks between merging this PR and publication of the article (or even afterwards; for something like a typo fix, approval is pretty trivial).

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+and maintain a webhook binary to handle admission requests. Also, admission
+webhooks are complex to operate. Each webhook must be deployed, monitored and
+have a well defined upgrade and rollback plan. To make matters worse, if a
+webhook times out or becomes unavailable, the Kubernetes control plan can become
```
Comment (JimBugwadia) [inline content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:27]: Kubernetes control plane

**Comment by jpbetz** [comment]
squashed

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+by cluster administrators.
+
+For example, the above admission policy can be modified to make it configurable:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "demo-policy.example.com"
+Spec:
```
Comment (jmcshane) [inline content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:92]:
```suggestion
spec:
```

**Comment by sftim** [comment]
Hmm, good point!

/lgtm cancel

File: content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md (hunk 11 above)
Comment (fsmunoz) [reply content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:92]: @jpbetz I think this is the only point left before being able to approve/merge.

Comment (sftim) [reply content/en/blog/_posts/2022-12-20-validating-admission-policies-alpha/index.md:92]: I can (and will) use admin access to accept this change.

**Comment by sftim** [comment]
Thanks
/lgtm

**Pull request merged**